### PR TITLE
[NO JIRA]: Correct types for datepicker and dialog

### PR DESCRIPTION
**Pull request by olliecurtis**
This PR corrects some types that we had some issues with when we moved to TS.
- Corrects Datepicker input type to accept a ReactElement vs a shared prop across multiple components
- Improves readability of README for dialog to be clearer on the `onClose` prop
- Adds a runtime type check to ensure that the correct types and combinations are passed
- Makes the Dialog `onClose` prop optional for when dismissible is true

Remember to include the following changes:

- [ ] `README.md` (If you have created a new component)
- [x] Component `README.md`
- [x] Tests
- [ ] Storybook examples created/updated
- [ ] For breaking changes or deprecating components/properties, migration guides added to the description of the PR. If the guide has large changes, consider creating a new Markdown page inside the component's docs folder and link it here

### CODE REVIEW

File: packages/bpk-component-datepicker/src/BpkDatepicker.d.ts
Comment (olliecurtis) [inline packages/bpk-component-datepicker/src/BpkDatepicker.d.ts:1]: I am slightly worried about the changes in this file as this suddenly seems like a lot of changes?

Looks like a lot of reformatting of order, but I do see there is an added `Omit<` (L64) but also some addition of `globalThis.` which seems suspicious

Comment (anambl) [reply packages/bpk-component-datepicker/src/BpkDatepicker.d.ts:1]: Ohhh... that is a bit weird indeed. Not sure why it adds the `Omit` to the type of default props 😓 Although looking at the props https://github.com/Skyscanner/backpack/blob/main/packages/bpk-component-calendar/src/BpkCalendarContainer.tsx#L353-L388  I think it does make sense as it is overwriting those ommited props

For the `globalThis` I think we don't need to import JSX?

File: examples/bpk-animate-height/stories.js
Comment (anambl) [inline examples/bpk-animate-height/stories.js:1]: I think this is accidental? 😄

Comment (olliecurtis) [reply examples/bpk-animate-height/stories.js:1]: Whoops yes!

**Review by anambl — APPROVED** [review]

**Pull request merged**